flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java

```diff
@@ -0,0 +1,345 @@
+				}
+			} else {
+				KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+			}
+		} else {
+			KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+		}
+	}
+
+	/**
+	 * Response when running with leading JobManager.
+	 */
+	private void respondAsLeader(final ChannelHandlerContext ctx, final HttpRequest request, final Map<String, String> pathParams, final ActorGateway jobManager) throws Exception {
+		final String taskManagerID = pathParams.get(TaskManagersHandler.TASK_MANAGER_ID_KEY);
+
+		boolean fetch;
+		synchronized (lastRequestCompleted) {
+			if (!lastRequestCompleted.containsKey(taskManagerID)) {
+				lastRequestCompleted.put(taskManagerID, true);
+			}
+			fetch = lastRequestCompleted.get(taskManagerID);
+			lastRequestCompleted.put(taskManagerID, false);
+		}
+
+		//fetch taskmanager logs if no other process is currently doing it
+		if (fetch) {
+			//get taskmanager gateway
+			InstanceID instanceID = new InstanceID(StringUtils.hexStringToByte(taskManagerID));
+			Future<Object> future = jobManager.ask(new JobManagerMessages.RequestTaskManagerInstance(instanceID), timeout);
+			JobManagerMessages.TaskManagerInstance instance = (JobManagerMessages.TaskManagerInstance) Await.result(future, timeout);
+			Instance taskManager = instance.instance().get();
+
+			Future<Object> isBlobServiceDefined = taskManager.getActorGateway().ask(TaskManagerMessages.getIsBlobServiceDefined(), timeout);
+
+			if (!(Boolean) Await.result(isBlobServiceDefined, timeout)) {
+				display(ctx, request, "BlobService unavailable, cannot upload TaskManager logs.");
+				return;
+			}
+
+			//send log request to taskmanager
+			Future<Object> blobKeyFuture = taskManager.getActorGateway().ask(serveLogFile ? TaskManagerMessages.getRequestTaskManagerLog() : TaskManagerMessages.getRequestTaskManagerStdout(), timeout);
+			blobKeyFuture.onComplete(new OnComplete<Object>() {
+				@Override
+				public void onComplete(Throwable failure, Object success) throws Throwable {
+					try {
+						BlobKey blobKey = (BlobKey) success;
+
+						//delete previous log file, if it is different than the current one
+						if ((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).containsKey(taskManagerID)) {
+							if (!blobKey.equals((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).get(taskManagerID))) {
+								jobManager.tell(JobManagerMessages.getDeleteTaskManagerLog((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).remove(taskManagerID)));
+								(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+							}
+						} else {
+							(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+						}
+
+						//send blobkey to jobmanager
+						Future<Object> logPathFuture = jobManager.ask(JobManagerMessages.getRequestTaskManagerLog(blobKey), timeout);
```

Comment:
I don't like that we request the file path here. This more or less means that the handler must have access to the filesystem as the JM has. In the long run, we want to disentangle the `WebRuntimeMonitor` from the `JobManager` so that both components are independent of each other. I would rather prefer that the `WebRuntimeMonitor` has its own `BlobService` which it uses to request the log file given the `BlobKey` received from the TaskManager. In the first implementation, this `BlobService` could simply be the `BlobServer` of the JM. By doing it this way, we would create an abstraction which would make it easier to separate the `WebRuntimeMonitor` from the `JobManager`.